integration_tests/integration_test.go

```diff
@@ -260,15 +261,200 @@ var _ = Describe("Tablecloth HTTP listener", func() {
+			})
+		})
+
+		Context("the new server exits shortly after starting", func() {
+			It("should continue running the old server", func() {
+				resp, err := http.Get("http://127.0.0.1:8081/")
+				Expect(err).NotTo(HaveOccurred())
+				Expect(resp.StatusCode).To(Equal(200))
+				firstBody, _ := ioutil.ReadAll(resp.Body)
+				resp.Body.Close()
+
+				Expect(os.Remove(cwd + "/test_servers/current")).To(Succeed())
+				Expect(os.Symlink(cwd+"/test_servers/errorer", cwd+"/test_servers/current")).To(Succeed())
+
+				withSilentOutput(func() {
+					reloadServer(serverCmd)
+				})
+
+				resp, err = http.Get("http://127.0.0.1:8081/")
+				Expect(err).NotTo(HaveOccurred())
+				Expect(resp.StatusCode).To(Equal(200))
+				newBody, _ := ioutil.ReadAll(resp.Body)
+				resp.Body.Close()
+
+				Expect(newBody).To(Equal(firstBody))
+			})
+
+			if canReadProcessFds() {
+				It("should not leak file descriptors when reloading fails", func() {
+					resp, err := http.Get("http://127.0.0.1:8081/")
+					Expect(err).NotTo(HaveOccurred())
+					Expect(resp.StatusCode).To(Equal(200))
+
+					initalFds := getProcessFds(serverCmd)
+
+					Expect(os.Remove(cwd + "/test_servers/current")).To(Succeed())
+					Expect(os.Symlink(cwd+"/test_servers/errorer", cwd+"/test_servers/current")).To(Succeed())
+
+					withSilentOutput(func() {
+						reloadServer(serverCmd)
+					})
+
+					resp, err = http.Get("http://127.0.0.1:8081/")
+					Expect(err).NotTo(HaveOccurred())
+					Expect(resp.StatusCode).To(Equal(200))
+
+					currentFds := getProcessFds(serverCmd)
+					Expect(currentFds).To(Equal(initalFds))
+				})
+			} else {
+				PIt("leaking file descriptors test requires /proc/<pid>/fd directories")
+			}
+
+			It("should successfully handle subsequent reload requests with a good server", func() {
+				resp, err := http.Get("http://127.0.0.1:8081/")
+				Expect(err).NotTo(HaveOccurred())
+				Expect(resp.StatusCode).To(Equal(200))
+
+				Expect(os.Remove(cwd + "/test_servers/current")).To(Succeed())
+				Expect(os.Symlink(cwd+"/test_servers/errorer", cwd+"/test_servers/current")).To(Succeed())
+
+				withSilentOutput(func() {
+					reloadServer(serverCmd)
+				})
+				resp, err = http.Get("http://127.0.0.1:8081/")
+				Expect(err).NotTo(HaveOccurred())
+				Expect(resp.StatusCode).To(Equal(200))
+
+				// Now point back at a good server
+				Expect(os.Remove(cwd + "/test_servers/current")).To(Succeed())
+				Expect(os.Symlink(cwd+"/test_servers/v2", cwd+"/test_servers/current")).To(Succeed())
+
+				reloadServer(serverCmd)
+
+				resp, err = http.Get("http://127.0.0.1:8081/")
+				Expect(err).NotTo(HaveOccurred())
+				Expect(resp.StatusCode).To(Equal(200))
+
+				body, _ := ioutil.ReadAll(resp.Body)
+				Expect(string(body)).To(ContainSubstring("Hello from v2"))
+			})
+		})
+	})
 })
 
+var serverOutputWriter = &struct{ io.Writer }{os.Stderr}
+
+func withSilentOutput(f func()) {
+	serverOutputWriter.Writer = ioutil.Discard
```

Comment:
That's interesting... I usually run `go test ./...`, which does not print stderr without the verbose flag, rather than simply `go test`, which does print stderr without the verbose flag:

I tested by removing `withSilentOutput`:

```bash
matt➜alext/tablecloth/integration_tests(handle_server_start_error✗)» go test .
ok  	github.com/alext/tablecloth/integration_tests	34.628s
```

...versus...

```bash
matt➜alext/tablecloth/integration_tests(handle_server_start_error✗)» go test
Running Suite: Tablecloth
=========================
Random Seed: 1430504487
Will run 13 of 16 specs

•••
------------------------------
P [PENDING]
Tablecloth HTTP listener
/Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:434
  Listening on a single port
  /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:112
    leaking file descriptors test requires /proc/<pid>/fd directories
    /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:110
------------------------------
••••••2015/05/01 19:21:48 [tablecloth] error starting new server, aborting reload: chdir /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/test_servers/current: no such file or directory
•
------------------------------
P [PENDING]
Tablecloth HTTP listener
/Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:434
  handling failures starting the new version
  /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:433
    the new server fails to start
    /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:358
      leaking file descriptors test requires /proc/<pid>/fd directories
      /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:325
------------------------------
2015/05/01 19:21:50 [tablecloth] error starting new server, aborting reload: chdir /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/test_servers/current: no such file or directory
•2015/05/01 19:21:54 [tablecloth] error starting new server, aborting reload: child no longer running after StartupDelay(100ms)
•
------------------------------
P [PENDING]
Tablecloth HTTP listener
/Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:434
  handling failures starting the new version
  /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:433
    the new server exits shortly after starting
    /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:432
      leaking file descriptors test requires /proc/<pid>/fd directories
      /Users/matt/go/src/github.com/alext/tablecloth/integration_tests/integration_test.go:403
------------------------------
2015/05/01 19:21:56 [tablecloth] error starting new server, aborting reload: child no longer running after StartupDelay(100ms)
•
Ran 13 of 16 Specs in 29.801 seconds
SUCCESS! -- 13 Passed | 0 Failed | 3 Pending | 0 Skipped PASS
ok  	github.com/alext/tablecloth/integration_tests	32.148s
```

Anyway, I think what you have is fine. :+1: